http-clients/netty-nio-client/src/test/java/software/amazon/awssdk/http/nio/netty/fault/ServerConnectivityErrorMessageTest.java

```diff
@@ -0,0 +1,356 @@
+import io.netty.handler.codec.http.HttpVersion;
+import io.netty.handler.codec.http.LastHttpContent;
+import io.netty.handler.ssl.SslContext;
+import io.netty.handler.ssl.SslContextBuilder;
+import io.netty.handler.ssl.SslHandler;
+import io.netty.handler.ssl.util.SelfSignedCertificate;
+import io.reactivex.Flowable;
+import java.nio.ByteBuffer;
+import java.time.Duration;
+import java.util.Arrays;
+import java.util.Collection;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.ExecutionException;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.TimeoutException;
+import javax.net.ssl.SSLEngine;
+import javax.net.ssl.SSLEngineResult;
+import javax.net.ssl.SSLException;
+import org.junit.After;
+import org.junit.Before;
+import org.junit.Test;
+import org.junit.runner.RunWith;
+import org.junit.runners.Parameterized;
+import org.reactivestreams.Publisher;
+import software.amazon.awssdk.http.EmptyPublisher;
+import software.amazon.awssdk.http.Protocol;
+import software.amazon.awssdk.http.SdkHttpFullRequest;
+import software.amazon.awssdk.http.SdkHttpMethod;
+import software.amazon.awssdk.http.SdkHttpResponse;
+import software.amazon.awssdk.http.async.AsyncExecuteRequest;
+import software.amazon.awssdk.http.async.SdkAsyncHttpClient;
+import software.amazon.awssdk.http.async.SdkAsyncHttpResponseHandler;
+import software.amazon.awssdk.http.nio.netty.NettyNioAsyncHttpClient;
+import software.amazon.awssdk.http.nio.netty.SdkEventLoopGroup;
+import software.amazon.awssdk.utils.AttributeMap;
+import software.amazon.awssdk.utils.Logger;
+
+@RunWith(Parameterized.class)
+public class ServerConnectivityErrorMessageTest {
+    private static final Logger LOGGER = Logger.loggerFor(ServerConnectivityErrorMessageTest.class);
+    private SdkAsyncHttpClient netty;
+    private Server server;
+
+    @Parameterized.Parameter
+    public TestCase testCase;
+
+    @Parameterized.Parameters(name = "{0}")
+    public static Collection<TestCase> testCases() {
+        return Arrays.asList(new TestCase(CloseTime.DURING_INIT, "The connection was closed during the request."),
+                             new TestCase(CloseTime.BEFORE_SSL_HANDSHAKE, "The connection was closed during the request."),
+                             new TestCase(CloseTime.DURING_SSL_HANDSHAKE, "The connection was closed during the request."),
+                             new TestCase(CloseTime.BEFORE_REQUEST_PAYLOAD, "The connection was closed during the request."),
+                             new TestCase(CloseTime.DURING_REQUEST_PAYLOAD, "The connection was closed during the request."),
+                             new TestCase(CloseTime.BEFORE_RESPONSE_HEADERS, "The connection was closed during the request."),
+                             new TestCase(CloseTime.BEFORE_RESPONSE_PAYLOAD, "Response had content-length"),
+                             new TestCase(CloseTime.DURING_RESPONSE_PAYLOAD, "Response had content-length"));
```

Comment:
Awesome test cases. Can we come up with more distinctive wording for the first 6 test cases, or out of scope of this PR?

Reply:
Do you have an example what you mean?

Reply:
Something like this (numbers are not aligned):

1. The connection was closed while it was being initialized.
2. The connection was closed during the TLS handshake process.
3. The connection was closed before it could be written to.
4. The connection was closed while sending the request body.
5. The connection was closed while waiting to receive a response.
6. The connection was closed while receiving the response body.

This type of wording could greatly help debugging certain failure scenarios.

Reply:
The test case already provides something similar through its toString. I feel like that should be enough without needing to maintain a string description.

![Screen Shot 2022-01-19 at 11 47 36 AM](https://user-images.githubusercontent.com/24903526/150203188-889c580c-c438-408a-bd6a-881c89646b91.png)

Reply:
I meant can we change the user-facing wording used in the actual exception?

Reply:
Gotcha, we can talk offline. I tried super hard to do that originally (that's why these test cases are structured like that), and it just doesn't work like that. The client thinks it's waiting for a response when the service hasn't even finished the SSL handshake.